### PR TITLE
[v18] feat(timepicker): add new component timepicker

### PR DESCRIPTION
**Pull request by skhamvon**
First step to add the new component timepicker

Initialization of component, basics tests and documentation

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build#1661.0 ✘  
* 
  * build ✘

**Comment by ovh-cds** [comment]
CDS Report build#1671.0 ✘  
* 
  * build ✘

**Comment by ovh-cds** [comment]
CDS Report build#1675.0 ■   
* 
  * build ■

**Comment by ovh-cds** [comment]
CDS Report build#1681.0 ■   
* 
  * build ■

**Comment by ovh-cds** [comment]
CDS Report build#1681.0 ■   
* 
  * build ■

File: packages/components/timepicker/package.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  "description": "ODS Timepicker component",
+  "license": "Apache-2.0",
+  "main": "dist/index.cjs.js",
+  "module": "dist/index.js",
+  "es2015": "dist/esm/index.js",
+  "es2017": "dist/esm/index.js",
+  "types": "dist/types/index.d.ts",
+  "collection": "dist/collection/collection-manifest.json",
+  "collection:main": "dist/collection/index.js",
+  "scripts": {
```
Comment (dpellier) [inline packages/components/timepicker/package.json:13]: please add the `lint:scss` command

File: packages/components/timepicker/src/components/osds-timepicker/constants/default-attributes.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { OdsTimepickerAttribute } from '../interfaces/attributes';
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/constants/default-attributes.ts:1]: `import type`

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/components/timepicker/package.json (hunk 1 above)
Comment (skhamvon) [reply packages/components/timepicker/package.json:13]: Added

File: packages/components/timepicker/src/components/osds-timepicker/constants/default-attributes.ts (hunk 2 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/constants/default-attributes.ts]: It's changed

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    return (
+      <Host>
+        <osds-input clearable={ clearable }
+                    disabled={ disabled }
+                    error={ error }
+                    inline={ inline }
+                    step={ withSeconds ? 1 : "" }
+                    type={ ODS_INPUT_TYPE.time }
+                    value={ value }
+                    {...{
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx:60]: no need for destructuring

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  @Prop({ reflect: true, mutable: true }) public value? = DEFAULT_ATTRIBUTE.value;
+
+  /** @see OdsTimepickerAttribute.withSeconds */
+  @Prop({ reflect: true }) public withSeconds? = DEFAULT_ATTRIBUTE.withSeconds;
+
+  @Watch('withSeconds')
+  checkSeconds(withSeconds: boolean) {
+    if(withSeconds === false && this.inputEl?.value.match(/:/g)?.length === 2 ){
+      const inputValue = this.inputEl.value.split(':');
+      this.inputEl.value = inputValue[0].concat(':', inputValue[1]);
```
Comment (dpellier) [inline packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx:37]:
```suggestion
      this.inputEl.value = `${inputValue[0]}:${inputValue[1]}`;
```

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx (hunk 3 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx]: Fixed

File: packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx (hunk 4 above)
Comment (skhamvon) [reply packages/components/timepicker/src/components/osds-timepicker/osds-timepicker.tsx]: Changed

**Review by dpellier — APPROVED** [review]

**Review by feoche — CHANGES_REQUESTED** [review]
Wrong documentation category in Storybook

**Review by astagnol — CHANGES_REQUESTED** [review]
- There's a font issue on Safari:
Capture d’écran 2023-11-13 à 15 06 14

- On safari, set inline to true, add a timezone -> timezones input is not displayed as expected:
Capture d’écran 2023-11-13 à 15 10 07
FYI, if you add a second timezone, first one is displayed as expected as default value.

- On Safari, clearable does not clear the input

- It would be nice to add a story or an example with preset timezones so users who are not used to Storybook would know optional timezones input exists. wdyt?

**Comment by feoche** [comment]
UI Review :
- Keyboard navigation does not permit to go to minutes (seconds)
- When adding timezones as attributes, Timezones select is broken

**Review by astagnol — CHANGES_REQUESTED** [review]

File: packages/components/src/timepicker/documentation/specifications/spec.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+* [**Interfaces**](#interfaces)
+* [**Classes**](#classes)
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/spec.md]: Removed

File: packages/storybook/stories/components/timepicker/3_demo.stories.ts [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+/* With Timezones Example */
+const TemplateTimezonesExample = (args: any) => html`
+  <osds-timepicker ...=${getTagAttributes({
+    ...args,
+  })}>
+  </osds-timepicker>
+`;
+export const TimezonesExample = TemplateTimezonesExample.bind({});
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/storybook/stories/components/timepicker/3_demo.stories.ts]: Done

File: packages/storybook/stories/components/timepicker/3_demo.stories.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+const defaultParam = {
+  timezones: {
+    category: 'Timezones',
+    defaultValue: [],
+  },
+};
+
+const exampleParam = {
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/storybook/stories/components/timepicker/3_demo.stories.ts:42]: Done

File: packages/storybook/stories/components/timepicker/3_demo.stories.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    category: 'General',
+    defaultValue: false,
+  },
+  error: {
+    category: 'General',
+    defaultValue: false,
+  },
+  inline: {
+    category: 'General',
+    defaultValue: false,
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/storybook/stories/components/timepicker/3_demo.stories.ts]: Changed

File: packages/components/src/timepicker/package.json [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    "test:spec:ci": "stencil test --config stencil.config.ts  --spec --ci --coverage",
+    "FIXME_test:e2e:ci:screenshot": "stencil test --config stencil.config.ts --e2e --ci --screenshot --passWithNoTests",
+    "FIXME_test:e2e:ci:screenshot:update": "stencil test --config stencil.config.ts --e2e --ci --screenshot --update-screenshot --passWithNoTests"
+  },
+  "dependencies": {
+    "@ovhcloud/ods-common-core": "17.0.0",
+    "@ovhcloud/ods-common-stencil": "17.0.0",
+    "@ovhcloud/ods-common-theming": "17.0.0",
+    "@ovhcloud/ods-component-input": "17.0.0",
+    "@ovhcloud/ods-component-select": "17.0.0"
```
Comment (dpellier) [inline packages/components/src/timepicker/package.json:37]: remove both ods-component-* dependencies

File: packages/components/src/timepicker/package.json [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  "devDependencies": {
+    "@ovhcloud/ods-common-scripts": "17.0.0",
```
Comment (dpellier) [inline packages/components/src/timepicker/package.json:40]: remove

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    if (withSeconds === true && value.match(/:/g)?.length === 1) {
+      return `${ value }:00`;
+    }
+
+    return '';
+  }
+
+  handleTimezones() :void {
+    const check = this.parseTimezones(this.getTimezonesList());
+    this.component.timezonesList = [...check] as ODS_TIMEZONE[] ;
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts:26]: no destructuring needed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        let parsedTimezones = JSON.parse(timezones);
+        parsedTimezones = parsedTimezones.filter(
+          (parsedTimezone: string) => ODS_TIMEZONES.find((timezone: string) => timezone === parsedTimezone.toUpperCase())).map((parsedTimezone: string) => parsedTimezone.toUpperCase());
+        return parsedTimezones;
+      } catch (err) {
+        this.component.logger.warn('[OsdsTimepicker] timezones string could not be parsed.');
+        return [];
+      }
+    }
+    return [...timezones];
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts:41]: no destructuring needed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  handleCurrentTimezone() : void {
+    if (!this.component.currentTimezone) {
+      const browserTimezone = new Date().getTimezoneOffset() / 60 * -1;
+      const parsedTimezone = browserTimezone >= 0 ? `+${browserTimezone}` : browserTimezone.toString();
+      this.component.currentTimezone = ODS_TIMEZONES.find((timezone) => (timezone as string).indexOf(parsedTimezone) > 1);
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts:48]: `> -1`

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker-tests.mdx
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker-tests.mdx:1]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker-properties.mdx
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker-properties.mdx:1]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker-methods.mdx
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker-methods.mdx:1]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker-events.mdx
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker-events.mdx:1]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker-contents.mdx
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker-contents.mdx:1]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker.mdx [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import {Description} from '@storybook/addon-docs';
+import Specs from './spec.md';
+import SpecsTimepickerContents from './specifications-timepicker-contents.mdx';
+import SpecsTimepickerTests from './specifications-timepicker-tests.mdx';
+
+<Description>{Specs}</Description>
+
+## Contents
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker.mdx]: removed

File: packages/components/src/timepicker/documentation/specifications/specifications-timepicker.mdx [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+<SpecsTimepickerContents />
+
+## Tests
```
[the remark this answers is not in the document]
Comment (skhamvon) [reply packages/components/src/timepicker/documentation/specifications/specifications-timepicker.mdx]: removed

File: packages/components/src/timepicker/package.json (hunk 9 above)
Comment (skhamvon) [reply packages/components/src/timepicker/package.json]: removed

File: packages/components/src/timepicker/package.json (hunk 10 above)
Comment (skhamvon) [reply packages/components/src/timepicker/package.json]: removed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 11 above)
Comment (skhamvon) [reply packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts]: changed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 12 above)
Comment (skhamvon) [reply packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts]: changed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 13 above)
Comment (skhamvon) [reply packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts]: Fixed

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+    if (withSeconds === true && value.match(/:/g)?.length === 1) {
+      return `${ value }:00`;
+    }
+
+    return '';
+  }
+
+  handleTimezones() :void {
+    const check = this.parseTimezones(this.getTimezonesList());
+    this.component.timezonesList = check as ODS_TIMEZONE[] ;
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts:26]: you shouldn't have to force type as it's the actual return type of `parseTimezones`

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+      }
+    }
+    return timezones as ODS_TIMEZONE[];
+  }
+
+  handleCurrentTimezone() : void {
+    if (!this.component.currentTimezone) {
+      const browserTimezone = new Date().getTimezoneOffset() / 60 * -1;
+      const parsedTimezone = browserTimezone >= 0 ? `+${browserTimezone}` : browserTimezone.toString();
+      this.component.currentTimezone = ODS_TIMEZONES.find((timezone) => (timezone as string).indexOf(parsedTimezone) > -1);
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts:48]: you shouldn't have to force `string` type on timezone

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by astagnol — APPROVED** [review]

File: packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts (hunk 17 above)
Comment (skhamvon) [reply packages/components/src/timepicker/src/components/osds-timepicker/core/controller.ts]: removed

**Review by dpellier — CHANGES_REQUESTED** [review]
There is some info missing regarding change events
For now we can listen to sub input `odsValueChange` event but it does not contains AM/PM info, plus this cause an issue with typing
Also TBD it may be nice to regroup timezone change and input change in one event?

It may be a good idea to wait for form common change and update it directly so that we release it uniformed directly.

File: packages/components/src/timepicker/src/components/osds-timepicker/interfaces/attributes.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+   */
+  error?: boolean,
+  /**
+   * Defines if the Timepicker should be displayed inline or not
+   */
+  inline?: boolean,
+  /**
+   * Defines the timezones available to show in the select
+   */
+  timezones?: ODS_TIMEZONE[] | ODS_TIMEZONES_PRESET | string,
```
Comment (dpellier) [inline packages/components/src/timepicker/src/components/osds-timepicker/interfaces/attributes.ts:28]: not sure `string` should be allowed

**Review by astagnol — CHANGES_REQUESTED** [review]
**Set a timezone:**
- When I set an hour and then, pick a timezone, the time input content is cleared => it should not do this since it could be frustrating for the user.

**Error state:**
- error style does not apply anymore (same on Input component). I've checked master and everything seems OK. => We need to dig about this issue and where it comes from. (when I set 'withSeconds' to true, for example, and then, set 'error' to true, error style is displayed but cannot be undone)
- btw, error style should apply to input field and timezone select (should be similar to Phone number behaviour)

**Storybook:**
- No CurrentTimezone control?
- CurrentTimezone = UTC+1  in DOM on osds-timepicker but this value doesn't not exist in timezones defaultValue control (so it isn't displayed as value in the Timezone Select)

**Comment by dpellier** [comment]
replaced by https://github.com/ovh/design-system/pull/381

**Pull request closed without merging**